HADOOP-18679. Add API for bulk/paged object deletion: Iceberg PoC (#10233)
apache/iceberg: docs/docs/aws.md

This means for any table manifests containing `s3a://` or `s3n://` file paths, `S3FileIO` is still able to read them.
This feature allows people to easily switch from S3A to `S3FileIO`.

> If you do wish to use S3A, see the "Using the Hadoop S3A Connector" section below.

Reviewer: ```suggestion
If you do wish to use S3A, see the "Using Hadoop S3A via HadoopFileIO" section below.
```
(see heading comment below)

I think you might be able to link to that heading to help with navigation.

--conf spark.sql.catalog.my_catalog.http-client.apache.max-connections=5
```

> ## Using the Hadoop S3A Connector

Reviewer: ```suggestion
Using Hadoop S3A via HadoopFileIO
```

This isn't really a connector so this heading is a little confusing.

The Apache Hadoop S3A Connector is an alternaive to S3FileIO. It:
* Uses an AWS v2 client qualified across a broad set of applications, including Apache Spark, Apache Hive, Apache HBase and more. This may lag the Iceberg artifacts.
* Contains detection and recovery for S3 failures beyond that in the AWS SDK -recovery added "one support call at a time".

> * Supports scatter/gather IO "Vector IO" for high-performance Parquet reads.

Reviewer: Is vectored IO supported in the Iceberg read path?  I wouldn't callout anything that isn't supported natively.

* Supports Amazon S3 Express One Zone storage, FIPS endpoints, Client-Side encryption S3 Access Points and S3 Access Grants.

> * Supports OpenSSL as an optional TLS transport layer -for tangible performance improvements over the JDK implementation.

Reviewer: ```suggestion
* Supports OpenSSL as an optional TLS transport layer for tangible performance improvements over the JDK implementation.
```

* Includes [auditing via the S3 Server Logs](https://hadoop.apache.org/docs/current/hadoop-aws/tools/hadoop-aws/auditing.html), which can be used to answer important questions such as "who deleted all the files?" and "which job is triggering throttling?".
* Collects [client-side statistics](https://apachecon.com/acasia2022/sessions/bigdata-1191.html) for identification of performance and connectivity issues.
* Note: it does not support S3 Dual Stack, S3 Acceleration or S3 Tags.

To use the S3A Connector, here are the instructions:

> 1. To store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3a://my-bucket/my-warehouse` 

Reviewer: ```suggestion
To store data using S3A, specify the `warehouse` catalog property to be an S3A path, e.g. `s3://my-bucket/my-warehouse`
```

Prefer using the `s3` scheme as it's more universal.

* Parquet should be configured to use the Vector IO API for parallelized data retrieval.

The recommended settings are listed below:

| Property                           | Recommended value | Description                                                                            |
|------------------------------------|-------------------|----------------------------------------------------------------------------------------|
| iceberg.hadoop.bulk.delete.enabled | true              | Iceberg to use Hadoop bulk delete API, where available                          |
| parquet.hadoop.vectored.io.enabled | true              | Use Vector IO in Parquet Reads, where available                                                    |
| fs.s3a.vectored.read.min.seek.size | 128K              | Threshold below which adjacent Vector IO ranges are coalesced into single GET requests |
| fs.s3a.experimental.input.fadvise      | parquet,vector,random,adaptive              | Preferred read policy when opening files.                                   |

```shell
spark-sql --conf spark.sql.catalog.my_catalog=org.apache.iceberg.spark.SparkCatalog \

>     --conf spark.sql.catalog.my_catalog.warehouse=s3a://my-bucket/my/key/prefix \

Reviewer: ```suggestion
--conf spark.sql.catalog.my_catalog.warehouse=s3://my-bucket/my/key/prefix
```
